### PR TITLE
Support ns manage donot need restart pod

### LINKED ISSUE #555: Do not need to restard service when executing ns management
**What would you like to be added**:

**Why is this needed**:
When Kmesh is deployed, if we label a namespace to let Kmesh manage the traffic of pods in the namespace there is currently a need to restart the service. Actually, we do not want to restart the service.

### PR DESCRIPTION
**Pull request by weli-l**
**What type of PR is this?**
support ns manage donot need restart pod

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #555 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/controller/manage/manage_controller.go [hunk 1]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+		log.Errorf("failed to enable Kmesh manage")
+		return
+	}
+	queue.AddRateLimited(QueueItem{podName: pod.Name, podNs: pod.Namespace, action: ActionAddAnnotation})
+}
+
+func disableKmeshManage(pod *corev1.Pod, queue workqueue.RateLimitingInterface, security *kmeshsecurity.SecretManager) {
+	sendCertRequest(security, pod, kmeshsecurity.DELETE)
+	if !isPodReady(pod) {
+		log.Debugf("Pod add event: %s/%s is not ready, skipping Kmesh manage enable", pod.GetNamespace(), pod.GetName())
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:246]: wrong log

File: pkg/controller/manage/manage_controller.go [hunk 2]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+	}
+
+	namespace, err := namespaceLister.Get(pod.Namespace)
+	if err != nil {
+		log.Errorf("failed to get pod namespace %s: %v", pod.Namespace, err)
+		return
+	}
+
+	if !utils.ShouldEnroll(pod, namespace) {
+		if pod.Annotations[constants.KmeshRedirectionAnnotation] != "enabled" {
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:144]: this is not right, should be ==

Comment (hzxuzhonghu) [reply pkg/controller/manage/manage_controller.go:144]: please add ut coverage

File: pkg/controller/manage/manage_controller.go [hunk 3]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+
+	if pod.Annotations[constants.KmeshRedirectionAnnotation] == "enabled" {
+		log.Infof("%s/%s: Pod managed by Kmesh is deleted", pod.GetNamespace(), pod.GetName())
+		sendCertRequest(security, pod, kmeshsecurity.DELETE)
+		// We donot need to do handleKmeshManage for delete, because we may have no change to execute a cmd in pod net ns.
+		// And we have done this in kmesh-cni
+	}
+}
+
+func handleNamespaceAddFunc(Obj interface{}, podLister v1.PodLister, queue workqueue.RateLimitingInterface, security *kmeshsecurity.SecretManager) {
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:200]: please change the function param to small cased, this is  golang idiomatic

File: pkg/controller/manage/manage_controller.go [hunk 4]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+	NS := Obj.(*corev1.Namespace)
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:201]: small cased

File: pkg/controller/manage/manage_controller.go [hunk 5]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+	}
+
+	if len(pods) == 0 {
+		clientset, err := utils.GetK8sclient()
+		if err != nil {
+			log.Errorf("Error getting Kubernetes client: %v", err)
+			return
+		}
+
+		podList, err := clientset.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{})
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:272]: We donot need this, if the pod lister has no pod, it means no pod event has received. We can rely on pod informer to trigger manage later,

File: pkg/controller/manage/manage_controller.go [hunk 6]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+		if err != nil {
+			log.Errorf("Error listing pods from API server for namespace %s: %v", namespace, err)
+			return
+		}
+		for i := range podList.Items {
+			pod := &podList.Items[i]
+			enableKmeshManage(pod, queue, security)
+		}
+	} else {
+		for _, pod := range pods {
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:282]: This is not right, please filter out only node in same node.  You can set list options podLister.Pods(namespace).List(labels.Everything())

File: pkg/controller/manage/manage_controller.go [hunk 7]
```diff
@@ -203,6 +127,176 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+			enableKmeshManage(pod, queue, security)
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:283]: This is not quite right, we can double calling this to a same pod

File: pkg/controller/manage/manage_controller.go (hunk 2 above)
Comment (weli-l) [reply pkg/controller/manage/manage_controller.go]: > please add ut coverage

added

File: pkg/controller/manage/manage_controller.go (hunk 3 above)
Comment (weli-l) [reply pkg/controller/manage/manage_controller.go]: > please change the function param to small cased, this is golang idiomatic

done

File: pkg/controller/manage/manage_controller.go (hunk 6 above)
Comment (weli-l) [reply pkg/controller/manage/manage_controller.go:282]: > This is not right, please filter out only node in same node. You can set list options podLister.Pods(namespace).List(labels.Everything())

I have set this above

File: pkg/controller/manage/manage_controller.go [hunk 8]
```diff
@@ -90,107 +91,29 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
-			}
+			handlePodDeleteFunc(obj, security)
 		},
 	}); err != nil {
 		return nil, fmt.Errorf("failed to add event handler to podInformer: %v", err)
 	}
 
+	if _, err := namespaceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(Obj interface{}) {
+			handleNamespaceAddFunc(Obj, podLister, queue, security)
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:108]: I think we donot need to handle ns add event. As if there are pods in the ns, we must receive pod add event alongside

File: pkg/controller/manage/manage_controller.go [hunk 9]
```diff
@@ -203,6 +126,164 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+
+	namespace, err := namespaceLister.Get(pod.Namespace)
+	if err != nil {
+		log.Errorf("failed to get pod namespace %s: %v", pod.Namespace, err)
+		return
+	}
+
+	if !utils.ShouldEnroll(pod, namespace) {
+		if pod.Annotations[constants.KmeshRedirectionAnnotation] == "enabled" {
+			disableKmeshManage(pod, queue, security)
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:144]: This branch is not tested

File: pkg/controller/manage/manage_controller.go [hunk 10]
```diff
@@ -203,6 +126,164 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+	}
+
+	namespace, err := namespaceLister.Get(newPod.Namespace)
+	if err != nil {
+		log.Errorf("failed to get pod namespace %s: %v", newPod.Namespace, err)
+		return
+	}
+
+	// enable kmesh manage
+	if oldPod.Annotations[constants.KmeshRedirectionAnnotation] != "enabled" && newPod.Annotations[constants.KmeshRedirectionAnnotation] != "enabled" && utils.ShouldEnroll(newPod, namespace) {
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:166]: I feel we can check 

`if newPod.Annotations[constants.KmeshRedirectionAnnotation] != "enabled" && utils.ShouldEnroll(newPod, namespace)`, not sure why we didnot

File: pkg/controller/manage/manage_controller.go [hunk 11]
```diff
@@ -203,6 +126,164 @@ func NewKmeshManageController(client kubernetes.Interface, security *kmeshsecuri
...
+		enableKmeshManage(newPod, queue, security)
+	}
+
+	// disable kmesh manage
+	if oldPod.Annotations[constants.KmeshRedirectionAnnotation] == "enabled" && newPod.Annotations[constants.KmeshRedirectionAnnotation] == "enabled" && !utils.ShouldEnroll(newPod, namespace) {
```
Comment (hzxuzhonghu) [inline pkg/controller/manage/manage_controller.go:171]:
```suggestion
	if newPod.Annotations[constants.KmeshRedirectionAnnotation] == "enabled" && !utils.ShouldEnroll(newPod, namespace) {
```

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**